client/exec.go

```diff
@@ -0,0 +1,312 @@
+		return nil, fmt.Errorf("opts.Stderr must be nil if opts.CombineStderr is true")
+	}
+
+	// Call the /v1/exec endpoint to start the command.
+	var timeoutStr string
+	if opts.Timeout != 0 {
+		timeoutStr = opts.Timeout.String()
+	}
+	payload := execPayload{
+		Command:       opts.Command,
+		Environment:   opts.Environment,
+		WorkingDir:    opts.WorkingDir,
+		Timeout:       timeoutStr,
+		UserID:        opts.UserID,
+		User:          opts.User,
+		GroupID:       opts.GroupID,
+		Group:         opts.Group,
+		UseTerminal:   opts.UseTerminal,
+		CombineStderr: opts.CombineStderr,
+		Width:         opts.Width,
+		Height:        opts.Height,
+	}
+	var body bytes.Buffer
+	err := json.NewEncoder(&body).Encode(&payload)
+	if err != nil {
+		return nil, err
+	}
+	headers := map[string]string{
+		"Content-Type": "application/json",
+	}
+	resultBytes, changeID, err := client.doAsyncFull("POST", "/v1/exec", nil, headers, &body)
+	if err != nil {
+		return nil, err
+	}
+	var result execResult
+	err = json.Unmarshal(resultBytes, &result)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check that we have the websocket IDs we expect.
+	wsIDs := result.WebsocketIDs
+	if wsIDs["control"] == "" {
+		return nil, fmt.Errorf(`response did not include "control" websocket`)
+	}
+	if wsIDs["io"] == "" {
+		return nil, fmt.Errorf(`response did not include "io" websocket`)
+	}
+	if !opts.CombineStderr && wsIDs["stderr"] == "" {
+		return nil, fmt.Errorf(`response did not include "stderr" websocket`)
+	}
+
+	// Connect to the "control" websocket.
+	controlConn, err := client.getChangeWebsocket(changeID, wsIDs["control"])
+	if err != nil {
+		return nil, err
+	}
+
+	// Forward stdin and stdout.
+	ioConn, err := client.getChangeWebsocket(changeID, wsIDs["io"])
+	if err != nil {
+		return nil, err
+	}
+	stdinDone := wsutil.WebsocketSendStream(ioConn, stdin, -1)
+	stdoutDone := wsutil.WebsocketRecvStream(stdout, ioConn)
+
+	// Handle stderr separately if needed.
+	var stderrConn *websocket.Conn
+	var stderrDone chan bool
+	if !opts.CombineStderr {
+		stderrConn, err = client.getChangeWebsocket(changeID, wsIDs["stderr"])
+		if err != nil {
+			return nil, err
+		}
+		stderrDone = wsutil.WebsocketRecvStream(stderr, stderrConn)
+	}
+
+	// Fire up a goroutine to wait for everything to be done.
+	writesDone := make(chan struct{})
+	go func() {
+		// Wait till the WebsocketRecvStream goroutines are done writing to
+		// stdout and stderr. This happens when EOF is signalled or websocket
+		// is closed.
+		<-stdoutDone
+		if stderrDone != nil {
+			<-stderrDone
+		}
+
+		// Empty the stdin channel, but don't block on it as stdin may be
+		// stuck in Read. This is due to the somewhat poor design of
+		// WebsocketSendStream, which writes to an unbuffered channel
+		// (instead of closing it or using a buffered channel). We'd rather
+		// not modify that package much, so it's closer to the LXD code.
+		go func() {
+			<-stdinDone // happens when reading stdin returns EOF
```

Comment:
The poor design described is indeed unfortunate, as we'll need to watch out for bugs in third-party logic that could create problems for us. Is there a chance this line may never execute, or is it guaranteed with something like a `defer` that this will always necessarily take place, even in error cases?  If it's not guaranteed, this will create goroutine leaks. If it is guaranteed, why are we forking off a new goroutine here?

Reply:
Yeah, I decided just to modify their library to close the channel instead (with a comment so it's easy to find mods later). Hopefully at some point we can create a library that's used by both projects with these quirks fixed.

The line would never have executed in the case of a bad `Read` provided by the user via the `Stdin` option that was blocked/stuck forever. If this happens now it'll still be a goroutine leak, but not two leaking. Hopefully passing in a bad / non-cancelling reader is not common. In any case, it's better than it was.

Reply:
Agreed on the update. It also feels like a Read that is stuck is indeed a problem on itself, but my main concern was actually that a Read that errors out of time doesn't suddenly make the underlying function return early and without closing the channel, which would make us leak a goroutine on every command executed.